### PR TITLE
[NET-10985] Fix bug where imagePullSecrets were not set up for Gateways

### LINKED ISSUE #4312: CR Api Gateway - deployment service account missing imagePullSecrets
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request. Searching for pre-existing feature requests helps us consolidate datapoints for identical requirements into a single place, thank you!
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.

---

### Overview of the Issue

CR api gateway deployment pod can't pull its consul-dataplane docker image. This is because when we deployed deploy CR with kind: Gateway, the correct deployment will be created with the correct pointing to serviceAccount, but unfortunately, this serviceAccount named `<gateway-name>-gateway` does not contain imagePullSecret in its definition.

### Reproduction Steps

1. Docker pull official images and add own tags
2. Docker push official images with own tags to private registry (Azure container registry) which is secured (ImagePullSecret is required)
3. Add secret on k8s -  myregistry.azurecr.io-access (kubernetes.io/dockerconfigjson) - is valid
4. Download official consul helm chart 1.5.3 (Consul 1.19.2)
5. Set in values yaml:

```
global:
  datacenter: mycenter
  name: consul
  image: myregistry.azurecr.io/repo/release/consul:1.19.2
  imageK8S: myregistry.azurecr.io/repo/release/consul-k8s-control-plane:1.5.3
  imageConsulDataplane: myregistry.azurecr.io/repo/release/consul-dataplane:1.5.3
  imagePullSecrets:
    - name: myregistry.azurecr.io-access 
``` 
  6. Install this chart on k8s 
  7. Ad own CR for consul Api Gateway

```
apiVersion: gateway.networking.k8s.io/v1beta1
kind: Gateway
metadata:
  name: api-gateway
  namespace: consul
spec:
  gatewayClassName: consul
  listeners:
  ...
```

 command to find created object with bad definition:

` kubectl get serviceAccount my-api-gateway -n ns`

Invalid serviceAccount definition:

```
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    component: api-gateway
    gateway.consul.hashicorp.com/created: "1725890955"
    gateway.consul.hashicorp.com/managed: "true"
    gateway.consul.hashicorp.com/name: int-mesh-gateway
    gateway.consul.hashicorp.com/namespace: test
  name: my-api-gateway
  namespace: test
  ownerReferences:
  - apiVersion: gateway.networking.k8s.io/v1beta1
    blockOwnerDeletion: true
    controller: true
    kind: Gateway
    name: my-api-gateway   
```

### Logs

 Logs in 1.19.2 version:

```
Events:
  Type     Reason     Age                   From               Message
  ----     ------     ----                  ----               -------
  Normal   Scheduled  14m                   default-scheduler  Successfully assigned spoc/my-api-gateway-58c4f65b9c-82ktg to aks-nodepool1-myVM
  Normal   Pulled     14m                   kubelet            Pulling image "myregistry.azurecr.io/repo/release/consul-k8s-control-plane:1.5.3" 
  Normal   Created    14m                   kubelet            Created container consul-connect-inject-init
  Normal   Started    14m                   kubelet            Started container consul-connect-inject-init
  Normal   Pulling    13m (x4 over 14m)     kubelet            Pulling image "myregistry.azurecr.io/repo/release/consul-dataplane:1.5.3"
  Warning  Failed     13m (x4 over 14m)     kubelet            Failed to pull image "myregistry.azurecr.io/repo/release/consul-dataplane:1.5.3": failed to pull and unpack image "myregistry.azurecr.io/repo/release/consul-dataplane:1.5.3": failed to resolve reference "myregistry.azurecr.io/repo/release/consul-dataplane:1.5.3": failed to authorize: failed to fetch anonymous token: unexpected status from GET request to https://myregistry.azurecr.io/oauth2/token?scope=repository%3Arelease%2Fconsul-dataplane%3Apull&service=myregistry.azurecr.io: 401 Unauthorized
  Warning  Failed     13m (x4 over 14m)     kubelet            Error: ErrImagePull
  Warning  Failed     13m (x5 over 14m)     kubelet            Error: ImagePullBackOff
  Normal   BackOff    4m39s (x42 over 14m)  kubelet            Back-off pulling image "myregistry.azurecr.io/repo/release/consul-dataplane:1.5.3"
```

### Expected behavior

The ServiceAccount created by the controller adding objects after detecting CR gateway will contain the correct imagePullSecret defined in the helm chart, in the global section

### Environment details

- `consul-k8s` version: 1.5.3
- `values.yaml` used to deploy the helm chart:

```
global:
  datacenter: mycenter
  name: consul
  image: myregistry.azurecr.io/repo/release/consul:1.19.2
  imageK8S: myregistry.azurecr.io/repo/release/consul-k8s-control-plane:1.5.3
  imageConsulDataplane: myregistry.azurecr.io/repo/release/consul-dataplane:1.5.3
  imagePullSecrets:
    - name: myregistry.azurecr.io-access 
```

### Additional Context

I think the definition below is wrong:

https://github.com/hashicorp/consul-k8s/blob/v1.5.3/control-plane/gateways/serviceaccount.go

The ServiceAccount definition is missing the imagePullSecret if it was added in the helm chart under  `global:`

In version 1.19.2 (consul-k8s1.5.3) you fixed the bug only for pulling the consul-k8s-control-plane image, which is needed to init container api gateway pod in #3862 but not for api gateway deployment image - consul-dataplane.

By the way, check this for other CR creation cases where serviceAccount from imagePullSecret is required.

### PR DESCRIPTION
**Pull request by nathancoleman**
Fixes #4312

> [!NOTE]
> In order for the pull secrets to work for a `Gateway`, they must be available in any namespace that a `Gateway` is deployed to. This is already the case with injected mesh sidecars if you, for example, consume consul-dataplane from a private image registry, so I have not made any special accomadations for `Gateways`.

### Changes proposed in this PR ###  
Plumb `global.imagePullSecrets` onto the `ServiceAccount` created for each `Gateway`

### How I've tested this PR ###
1. Created a private registry on DockerHub for consul-dataplane, which is used by the gateway's `Deployment`
    ```shell
    docker pull hashicorp/consul-dataplane:1.5.3
    docker tag hashicorp/consul-dataplane:1.5.3 <your_dockerhub_username>/consul-dataplane:1.5.3
    docker login
    docker push <your_dockerhub_username>/consul-dataplane:1.5.3
    ```
1. [Created an image pull secret for DockerHub in my K8s cluster](https://kubernetes.io/docs/tasks/configure-pod-container/pull-image-private-registry/#create-a-secret-by-providing-credentials-on-the-command-line)
1. Set `global.imageConsulDataplane` to the private registry version
1. Install using this version of the Helm chart and this build of consul-k8s-control-plane
    <details>
    <summary>values.yaml</summary>

    ```yaml
    global:
      name: consul
      datacenter: dc1
      imageConsulDataplane: docker.io/<your_dockerhub_username>/consul-dataplane:1.5.3
      imageK8S: consul-k8s-control-plane:local
      imagePullSecrets:
      - name: regcred
      tls:
        enabled: true
        enableAutoEncrypt: true
      acls:
        manageSystemACLs: true
    connectInject:
      enabled: true
    ```
    </details>

    ```shell
    kind create cluster
    make dev-docker && kind load docker-image consul-k8s-control-plane:local
    helm upgrade --install consul /path/to/consul-k8s/charts/consul --namespace consul --create-namespace --values ./values.yaml
    ```

### How I expect reviewers to test this PR ###
See above

### Checklist ###
- [ ] Tests added
- [x] [CHANGELOG entry added](https://github.com/hashicorp/consul-k8s/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

**Comment by pawellegowski89** [comment]
Looks ok from my side :)

**Comment by nathancoleman** [comment]
One thing that occurred to me while testing this yesterday is that the pull secrets will need to be available in any namespace where you want to deploy a `Gateway`; however, this is no different from deploying mesh sidecars today if you're deploying with a consul-dataplane from a private image repo, for example.

File: charts/consul/templates/connect-inject-configmap.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{{- if .Values.connectInject.enabled }}
```
Comment (jm96441n) [inline charts/consul/templates/connect-inject-configmap.yaml:1]: do we need to create this is `imagePullSecrets` is not set?

File: control-plane/api-gateway/gatekeeper/init.go [hunk 2]
```diff
@@ -36,9 +36,9 @@ type initContainerCommandData struct {
 	LogJSON  bool
 }
 
-// containerInit returns the init container spec for connect-init that polls for the service and the connect proxy service to be registered
+// initContainer returns the init container spec for connect-init that polls for the service and the connect proxy service to be registered
 // so that it can save the proxy service id to the shared volume and boostrap Envoy with the proxy-id.
-func (g Gatekeeper) initContainer(config common.HelmConfig, name, namespace string) (corev1.Container, error) {
+func (g *Gatekeeper) initContainer(config common.HelmConfig, name, namespace string) (corev1.Container, error) {
```
Comment (jm96441n) [inline control-plane/api-gateway/gatekeeper/init.go:41]: was changing this a pointer receiver to keep it consistent with the rest of the methods for the `Gatekeeper`?

File: charts/consul/templates/connect-inject-configmap.yaml (hunk 1 above)
Comment (nathancoleman) [reply charts/consul/templates/connect-inject-configmap.yaml:1]: Good call, I also need to gate the `config-file` arg passed to the connect injector `Deployment`

File: control-plane/api-gateway/gatekeeper/init.go [hunk 3]
```diff
@@ -36,9 +36,9 @@ type initContainerCommandData struct {
 	LogJSON  bool
 }
 
-// containerInit returns the init container spec for connect-init that polls for the service and the connect proxy service to be registered
+// initContainer returns the init container spec for connect-init that polls for the service and the connect proxy service to be registered
```
Comment (jm96441n) [inline control-plane/api-gateway/gatekeeper/init.go:39]: :+1:

File: charts/consul/templates/connect-inject-configmap.yaml (hunk 1 above)
Comment (nathancoleman) [reply charts/consul/templates/connect-inject-configmap.yaml:1]: It makes sense to gate this for now; however, in the future I am expecting this to be the place we put structured config values like this one that don't fit neatly into a flag value

File: control-plane/api-gateway/gatekeeper/serviceaccount.go [hunk 4]
```diff
@@ -7,18 +7,18 @@ import (
...
+	"k8s.io/apimachinery/pkg/types"
 	ctrl "sigs.k8s.io/controller-runtime"
+	gwv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+
+	"github.com/hashicorp/consul-k8s/control-plane/api-gateway/common"
 )
 
 func (g *Gatekeeper) upsertServiceAccount(ctx context.Context, gateway gwv1beta1.Gateway, config common.HelmConfig) error {
-	if config.AuthMethod == "" && !config.EnableOpenShift {
+	if config.AuthMethod == "" && !config.EnableOpenShift && len(config.ImagePullSecrets) == 0 {
```
Comment (jm96441n) [inline control-plane/api-gateway/gatekeeper/serviceaccount.go:21]: can we add a comment here about why we can't delete the service account if there are image pull secrets or maybe extract this to a function that can describe it a bit?

File: charts/consul/templates/connect-inject-configmap.yaml (hunk 1 above)
Comment (jm96441n) [reply charts/consul/templates/connect-inject-configmap.yaml:1]: nvm, I see how we handle if the image pull secrets is empty in the code, always having it results in nicer code paths

File: control-plane/api-gateway/gatekeeper/serviceaccount.go (hunk 4 above)
Comment (nathancoleman) [reply control-plane/api-gateway/gatekeeper/serviceaccount.go:23]: added a comment

File: .changelog/4316.txt [hunk 5]
```diff
@@ -0,0 +1,5 @@
+```release-note:bug
+api-gateway: `global.imagePullSecrets` are now configured on the `ServiceAccount` for `Gateways`.
+
+Note: the referenced image pull Secret(s) must be present in any namespace that the `Gateway` is deployed to.
```
Comment (blake) [inline .changelog/4316.txt:4]:
```suggestion
Note: the referenced image pull Secret(s) must be present in the same namespace the `Gateway` is deployed to.
```

**Comment by pawellegowski89** [comment]
> @pawellegowski89 One thing that occurred to me while testing this yesterday is that the pull secrets will need to be available in any namespace where you want to deploy a `Gateway`; however, this is no different from deploying mesh sidecars today if you're deploying with a consul-dataplane from a private image repo, for example.

You are right @nathancoleman, but that is the responsibility, so to speak, on the preparation of ns, not on the chart consul itself.

**Review by missylbytes — APPROVED** [review]

**Pull request merged**